Create shareable stylelint package for lucca-front.

fbasmaison-lucca opened the pull request:
## Description

Add a shareable stylelint config package for Lucca Front.

-----

* It extends sensible rules provided by existing configurations.
* It lists deprecated variables and classes used in SCSS files, based on LF milestones in Github.
* A documentation on how to install, configure and use is provided.
* It is based on its [implementation in Cleemy](https://github.com/LuccaSA/Cleemy.Front.Lint/pull/26), and a previous workshop on the subject.

Commits address rules with errors individually, either by ignoring them for now, or fixing them.

Also rework npm scripts to provide fine grained aliases, as [provided at Cleemy](https://github.com/LuccaSA/Cleemy.Front/pull/556/files#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R11-R17).

### Next step

* Lint templates for deprecated with [html-eslint](https://html-eslint.org/docs/rules/no-restricted-attr-values).

### Examples

* Regexp illustrating a [combination of selectors](https://regex101.com/r/6yQzje/1)

-----

![LF features are marked as warnings or errors in a terminal](https://github.com/user-attachments/assets/bf99ead3-98fe-438e-9d7b-b8fb2b26fb42)

https://github.com/user-attachments/assets/301b1369-eae0-46ff-977b-2602a26e5dc5

Related to:

* https://linear.app/lucca/issue/CLF-63/stylelint-lint-deprecated-or-deleted-classes-and-components
* https://github.com/LuccaSA/Cleemy.Front.Lint/pull/26

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/16212/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/16475/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

fbasmaison-lucca commented:
@jeremie-lucca : J’ai intégré les changements dont je t’ai parlé dans cette PR.

On constate qu’il restait un `--spacings-S`, entre autres choses. :)

Cette [Issue](https://linear.app/lucca/issue/CLF-63/stylelint-lint-deprecated-or-deleted-classes-and-components) a plus de détails, notamment sur ce que j’imagine côté front. (cc @Supamiu)

Reparlons-en quand vous pourrez. :)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

fbasmaison-lucca commented on stylelint.config.js:
```diff
@@ -0,0 +1,36 @@
+			// TODO: Fix those rules after removing them.
+			files: ['**/*.scss'],
+			rules: {
+				// 'block-no-empty': null,
+				// 'font-family-no-missing-generic-family-keyword': null,
+				// 'no-descending-specificity': null,
+				// 'no-duplicate-selectors': null,
+				// 'property-no-vendor-prefix': null,
+				// 'scss/at-mixin-argumentless-call-parentheses': null,
+				// 'scss/at-mixin-pattern': null,
+				// 'scss/at-rule-conditional-no-parentheses': null,
+				// 'scss/at-rule-no-unknown': null,
+				// 'selector-not-notation': null,
```
The PR is on purpose, so that we can have a look at it together.

I commented these lines to enable default config rules for CSS and SCSS configs.

fbasmaison-lucca commented:
👩‍🍳

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28257/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28497/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28524/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28598/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28658/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28659/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28691/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28769/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28894/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

fbasmaison-lucca commented:
@Supamiu We moved on with SCSS linting today.

We should be good to go once you have pushed changes about getting LF versions. 🎉

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/28978/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29115/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29154/ui-diff)

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29254/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29287/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29299/ui-diff)

Supamiu commented:
LGTM but for some reason esbuild is now acting up

shajz commented on package.json:
```diff
@@ -30,8 +30,8 @@
     "build-storybook": "ng run ng:build-storybook",
-    "lint": "ng lint",
-    "lint:scss": "stylelint --config .stylelintrc packages/scss/src/**/*.scss",
+    "lint:ng": "ng lint",
+    "lint:scss": "stylelint packages/**/*.scss",
```
we could reuse the lint:fix scripts we have at Cleemy's ? :)

shajz commented on packages/stylelint-config/README.md:
````diff
@@ -0,0 +1,118 @@
+## Install
+
+```
+npm i --save-dev stylelint @lucca/stylelint-config
````
Shouldn't this be on its own repo? Why are we referencing this package in the main package.json file instead of the npm one?

shajz replied:
If this is the "home" for this package, maybe we want to publish it as well? In the Jenkinsfile, we could add a script for that!

Supamiu replied:
I think we wantr to publish it too yeah, so that it makes it easy to reuse it in other projects.

fbasmaison-lucca replied:
That’s the idea, yeah. I’m not quite versed in the exercise. If you guys can improve this. I’d be more than happy about it!

shajz replied:
Sooo we create a new repo for this?

fbasmaison-lucca replied:
I thought we would be publishing from here, as other packages are.

Those rules are mainly for LF so, if we create another repository, I understand we should create a more generic package for the stylelint config we want to use at Lucca, and another one here for rules specific to LF.

Is it worth it?

shajz replied:
Okay the name of the package confused me. Maybe we should it to `@lucca-front/stylelint-config` then? Or `@lucca/stylelint-config-prisme` to be more forward thinking :)

fbasmaison-lucca replied:
Dang! I actually did not notice `-front` was missing.

I’m not sure why Prisme would be involved. I understand it as the portal for the Design System, which includes LF, so yes for `@lucca-front/stylelint-config`!

Thanks @corentin. The package renamed with the last commit pushed.

fbasmaison-lucca commented:
> LGTM but for some reason esbuild is now acting up

Seems like this is creating issues on this repository, although it might work in other Projects:

```ts
		if (version === '0.0.0') {
			version = null;
		}
```

fbasmaison-lucca commented:
I’m making a fix to trigger a warning if the LF version is `null`.

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29529/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29538/ui-diff)

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29541/ui-diff)

fbasmaison-lucca commented on packages/stylelint-config/currentLFVersion.mjs:
I’ve just tested on ExpenseReports and `version` returns `null`.

I added it with `npm i --save-dev ../lucca-front/packages/stylelint-config` though. I don’t know if that makes any difference.

fbasmaison-lucca commented:
@Supamiu I added a link about html-eslint to the PR for a next step.

The issues we found during our meeting earlier should be fixed. (`files` and the regexp for `.button.mod-info`).

I updated the documentation with the workflow we followed.

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29621/ui-diff)

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/29622/ui-diff)

shajz commented:
First publish had to be done manually, here we go : https://www.npmjs.com/package/@lucca/stylelint-config-prisme

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/30127/ui-diff)

shajz commented on packages/stylelint-config/LFVersions.mjs:
```diff
@@ -1,6 +1,6 @@
 import * as os from 'node:os';
 import { readFileSync, writeFileSync } from 'node:fs';
-import { join } from '@angular/compiler-cli';
+import { join } from 'path';
```
```suggestion
import { join } from 'node:path';
```

🤔 not sure

fbasmaison-lucca replied:
@Supamiu : Whaddayasay?

Supamiu replied:
It is indeed the way to go since few nodejs updates, I wasn't sure of it.

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/30356/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/30380/ui-diff)

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/30447/ui-diff)

fbasmaison-lucca commented:
<!-- linear:isThreadRoot -->

I updated this PR with deprecated selectors in 19.2.0.

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/31008/ui-diff)

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/31642/ui-diff)

fbasmaison-lucca commented:
Update after 19.2.4 release:

* Merge conflicts fixed.
* Remove [weird string](https://github.com/LuccaSA/lucca-front/pull/2745/commits/0a12d61900956c0204f5215885f022d56a916613).
* Run `npm run stylelint:fix`.

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

CCNET-iLucca commented:
[Tests d'interfaces](https://qa.lucca.tech/31645/ui-diff)

fbasmaison-lucca commented:
Branch update after merge conflicts.

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook

LuccaIntegration commented:
:woman_cook: https://lucca-front.lucca.tech/PR-2745/storybook
